### PR TITLE
IGNITE-9549 control.sh add command to collect information on the dist…

### PR DESCRIPTION
**Pull request by a-polyakov**
…ribution of partitions and reset lost partitions

Signed-off-by: a-polyakov <polyakov.alexandr.alexandrovich@gmail.com>

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java [hunk 1]
```diff
@@ -782,6 +802,36 @@ private void legacyCacheIdleVerify(GridClient client, CacheArguments cacheArgs)
         }
     }
 
+    /**
+     * @param client Client.
+     * @param cacheArgs Cache args.
+     */
+    private void cacheDistribution(GridClient client, CacheArguments cacheArgs) throws GridClientException {
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java]: Not needed empty line.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java [hunk 2]
```diff
@@ -1563,6 +1613,41 @@ else if (CMD_SKIP_ZEROS.equals(nextArg))
...
+                        nextArg = nextArg("User attributes are expected to be separated by commas");
+
+                        Set<String> userAttributes = new HashSet();
+
+                        for (String userAttribute:nextArg.split(","))
+                            userAttributes.add(userAttribute.trim());
+
+                        cacheArgs.setUserAttributes(userAttributes);
+
+                        if (hasNextCacheArg())
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java]: Why not use ternary operator? as nextArg = hasNextCacheArg() ? nextArg("") : null;

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionGroup.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.List;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.internal.visor.VisorDataTransferObject;
+
+/**
+ * DTO for CacheDistributionTask, contains information about group
+ */
+public class CacheDistributionGroup extends VisorDataTransferObject {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionGroup.java]: Usually we have two constructors on DTO classes: default (with no arguments) and full - with all arguments (useful for testing and tooling).

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionNode.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.internal.visor.VisorDataTransferObject;
+
+/**
+ * DTO for CacheDistributionTask, contains information about node
+ */
+public class CacheDistributionNode extends VisorDataTransferObject {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionNode.java]: Usually we have two constructors on DTO classes: default (with no arguments) and full - with all arguments (useful for testing and tooling).

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionPartition.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import org.apache.ignite.internal.processors.cache.distributed.dht.GridDhtPartitionState;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.internal.visor.VisorDataTransferObject;
+
+/**
+ * DTO for CacheDistributionTask, contains information about partition
+ */
+public class CacheDistributionPartition extends VisorDataTransferObject {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionPartition.java]: Usually we have two constructors on DTO classes: default (with no arguments) and full - with all arguments (useful for testing and tooling).

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+        return new CacheDistributionTaskResult(infos, exceptions);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected VisorJob<CacheDistributionTaskArg, CacheDistributionNode> job(CacheDistributionTaskArg arg) {
+        return new CacheDistributionJob(arg, debug);
+    }
+
+    private static class CacheDistributionJob extends VisorJob<CacheDistributionTaskArg, CacheDistributionNode> {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: Missing Javadocs

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                        partition.setPartition(p);
+
+                        partition.setPrimary(assignment.primaryPartitions(node.id()).contains(p));
+
+                        partition.setState(part.state());
+
+                        partition.setUpdateCounter(part.updateCounter());
+
+                        partition.setSize(desc == null ? part.dataStore().fullSize() : part.dataStore().cacheSize(id));
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                        if (part == null)
+                            continue;
+
+                        final CacheDistributionPartition partition = new CacheDistributionPartition();
+
+                        group.getPartitions().add(partition);
+
+                        int p = part.id();
+                        partition.setPartition(p);
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                        partition.setPrimary(assignment.primaryPartitions(node.id()).contains(p));
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+                        partition.setState(part.state());
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                        partition.setUpdateCounter(part.updateCounter());
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                        partition.setSize(desc == null ? part.dataStore().fullSize() : part.dataStore().cacheSize(id));
+
+                    }
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+                }
+
+                return info;
+
+            }
+            catch (Exception e) {
+                throw new IgniteException(e);
+            }
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTask.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskArg.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+     * Default constructor.
+     */
+    public CacheDistributionTaskArg() {
+        // No-op.
+    }
+
+    /**
+     * @param caches Caches.
+     */
+    public CacheDistributionTaskArg(Set<String> caches) {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskArg.java]: How about  userAttributes in this constructor ?

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+                    row.setPrimary(partition.isPrimary());
+                    row.setState(partition.getState());
+                    row.setUpdateCounter(partition.getUpdateCounter());
+                    row.setSize(partition.getSize());
+                    row.setAddresses(node.getAddresses());
+                    row.setUserAttributes(node.getUserAttributes());
+
+                    rows.add(row);
+                }
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+            }
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+        out.println("[groupId,partition,nodeId,primary,state,updateCounter,partitionSize,nodeAddresses"+userAttributesName+"]");
+
+        int oldGroupId = 0;
+
+        for (Row row : rows) {
+            if (oldGroupId != row.groupId) {
+                out.println("[next group: id=" + row.groupId + ", name=" + row.groupName + ']');
+
+                oldGroupId = row.getGroupId();
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+            }
+
+            row.print(out);
+
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: not needed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 19]
```diff
@@ -0,0 +1,344 @@
...
+        nodeResultList = U.readList(in);
+        exceptions = U.readMap(in);
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        return S.toString(CacheDistributionTaskResult.class, this);
+    }
+
+    private static class Row implements Comparable {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: Why this class embeddbed into  CacheDistributionTaskResult?

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 20]
```diff
@@ -0,0 +1,344 @@
...
+
+                if (result == 0)
+                    result = nodeId.compareTo(other.nodeId);
+
+            }
+
+            return result;
+        }
+
+        public void print(PrintStream out) {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: Missing javadocs.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+
+            }
+
+        }
+
+        rows.sort(null);
+
+
+        StringBuilder userAttributesName=new StringBuilder();
+        if (!rows.isEmpty() && rows.get(0).userAttributes!=null){
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: missed empty line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/reset_lost_partitions/CacheResetLostPartitionsTask.java [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.visor.VisorJob;
+import org.apache.ignite.internal.visor.VisorOneNodeTask;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ *
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/reset_lost_partitions/CacheResetLostPartitionsTask.java]: Missing javadocs text

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/reset_lost_partitions/CacheResetLostPartitionsTask.java [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+public class CacheResetLostPartitionsTask extends VisorOneNodeTask<CacheResetLostPartitionsTaskArg, CacheResetLostPartitionsTaskResult> {
+    /** */
+    private static final long serialVersionUID = 0L;
+
+    /** {@inheritDoc} */
+    @Override protected VisorJob<CacheResetLostPartitionsTaskArg, CacheResetLostPartitionsTaskResult> job(CacheResetLostPartitionsTaskArg arg) {
+        return new CacheResetLostPartitionsJob(arg, debug);
+    }
+
+    private static class CacheResetLostPartitionsJob extends VisorJob<CacheResetLostPartitionsTaskArg, CacheResetLostPartitionsTaskResult> {
```
Comment (akuznetsov-gridgain) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/reset_lost_partitions/CacheResetLostPartitionsTask.java]: Missing javadocs

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java (hunk 19 above)
Comment (a-polyakov) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/cache/distribution/CacheDistributionTaskResult.java]: This class need for sort and format result

**Pull request closed without merging**